pkg/ebpf/events_derived.go

```diff
@@ -6,85 +6,33 @@ import (
 	"github.com/aquasecurity/tracee/types/trace"
 )
 
-type deriveFn func(EventDefinition) error
-
-func (t *Tracee) deriveEvent(event trace.Event) []trace.Event {
-	var deriveFns map[int32]deriveFn
-	var derivatives []trace.Event
-
-	eventID := int32(event.EventID)
-	switch eventID {
-	case CgroupMkdirEventID:
-		containerCreateFn := func(def EventDefinition) error {
-			cgroupId, err := getEventArgUint64Val(&event, "cgroup_id")
-			if err != nil {
-				return err
-			}
-
-			if info := t.containers.GetCgroupInfo(cgroupId); info.ContainerId != "" {
-				de := event
-				de.EventID = int(ContainerCreateEventID)
-				de.EventName = "container_create"
-				de.ReturnValue = 0
-				de.StackAddresses = make([]uint64, 1)
-				de.Args = []trace.Argument{
-					{ArgMeta: def.Params[0], Value: info.Runtime},
-					{ArgMeta: def.Params[1], Value: info.ContainerId},
-					{ArgMeta: def.Params[2], Value: info.Ctime.UnixNano()},
-				}
-				de.ArgsNum = 3
-
-				derivatives = append(derivatives, de)
-			}
-
-			return nil
-		}
-		deriveFns = map[int32]deriveFn{
-			ContainerCreateEventID: containerCreateFn,
-		}
-	case CgroupRmdirEventID:
-		containerRemoveFn := func(def EventDefinition) error {
-			cgroupId, err := getEventArgUint64Val(&event, "cgroup_id")
-			if err != nil {
-				return err
-			}
-
-			if info := t.containers.GetCgroupInfo(cgroupId); info.ContainerId != "" {
-				de := event
-				de.EventID = int(ContainerRemoveEventID)
-				de.EventName = "container_remove"
-				de.ReturnValue = 0
-				de.StackAddresses = make([]uint64, 1)
-				de.Args = []trace.Argument{
-					{ArgMeta: def.Params[0], Value: info.Runtime},
-					{ArgMeta: def.Params[1], Value: info.ContainerId},
-				}
-				de.ArgsNum = 2
-
-				derivatives = append(derivatives, de)
-			}
-
-			return nil
-		}
-		deriveFns = map[int32]deriveFn{
-			ContainerRemoveEventID: containerRemoveFn,
-		}
+func (t *Tracee) initEventDerivationMap() error {
+	t.eventDerivations = map[int32]map[int32]deriveFn{
+		CgroupMkdirEventID: {
+			ContainerCreateEventID: deriveContainerCreate(t),
+		},
+		CgroupRmdirEventID: {
+			ContainerRemoveEventID: deriveContainerRemoved(t),
+		},
 	}
 
+	return nil
```

Comment:
If you already made this step to initialize these functions here, you can also save some other runtime steps by adding the derived event "template", which will contain the event id, event name, return value, stack addresses, args, and args num of the event. Then, when deriving the event in runtime, there will be no need to get the event definition and set these fields again

Reply:
Indeed this is for optimizations purposes.
If you approve that there is no much of performance gain by doing that, then I'm ok with keeping it that way

Reply:
There might be a performance gain only if we had an event that is VERY frequently derived (this might be workload dependent as well), and even so a map access is a pretty cheap operation (notice that we access the derivation map for every event and it takes only 0.5% cpu, it is a small map however), so I think this can be considered a premature optimization.